tests/unit/lms/services/canvas_test.py

```diff
@@ -1,89 +1,282 @@
+    def test_if_theres_a_permissions_error_and_theres_no_matching_file_it_raises(
+        self, canvas_api_client, canvas_file_finder, public_url_for_file
+    ):
+        canvas_api_client.public_url.side_effect = CanvasAPIPermissionError
+        canvas_file_finder.find_matching_file_in_course.return_value = None
+
+        with pytest.raises(CanvasAPIPermissionError):
+            public_url_for_file(sentinel.file_id)
+
+    def test_it_doesnt_save_a_mapped_file_id_if_getting_that_files_public_url_fails(
+        self,
+        canvas_api_client,
+        canvas_file_finder,
+        module_item_configuration,
+        public_url_for_file,
+    ):
+        canvas_api_client.public_url.side_effect = CanvasAPIPermissionError
+        canvas_file_finder.find_matching_file_in_course.return_value = (
+            sentinel.found_file_id
+        )
+
+        with pytest.raises(CanvasAPIPermissionError):
+            public_url_for_file(sentinel.file_id)
+
+        assert (
+            module_item_configuration.get_canvas_mapped_file_id(sentinel.file_id)
+            == sentinel.file_id
+        )
+
+    @pytest.fixture
+    def module_item_configuration(self, db_session):
+        module_item_configuration = factories.ModuleItemConfiguration()
+        db_session.flush()
+        return module_item_configuration
+
+    @pytest.fixture
+    def with_mapped_file_id(self, module_item_configuration):
+        module_item_configuration.set_canvas_mapped_file_id(
+            sentinel.file_id, sentinel.mapped_file_id
+        )
 
+    @pytest.fixture
+    def public_url_for_file(self, canvas_service, module_item_configuration):
+        return partial(
+            canvas_service.public_url_for_file,
+            module_item_configuration,
+            course_id=sentinel.course_id,
+        )
 
-class TestFindMatchingFileInCourse:
-    def test_it_returns_the_id_if_theres_a_matching_file_in_the_course(
-        self, canvas_service, canvas_api_client
+    @pytest.fixture(autouse=True)
+    def CanvasFileFinder(self, patch):
+        return patch("lms.services.canvas.CanvasFileFinder")
+
+    @pytest.fixture
+    def canvas_file_finder(self, CanvasFileFinder):
+        return CanvasFileFinder.return_value
+
+
+class TestCanvasFileFinder:
+    def test_assert_file_in_course_doesnt_raise_if_the_file_is_in_the_course(
+        self, finder
+    ):
+        finder.assert_file_in_course(sentinel.course_id, "2")
+
+    def test_assert_file_in_course_raises_if_the_file_isnt_in_the_course(self, finder):
+        with pytest.raises(CanvasFileNotFoundInCourse):
+            finder.assert_file_in_course(sentinel.course_id, "4")
+
+    def test_find_matching_file_in_course_returns_the_matching_file_id(
+        self, finder, canvas_api_client, file_service
     ):
         # The file dict from the Canvas API that we expect the search to match.
         matching_file_dict = canvas_api_client.list_files.return_value[1]
-
```

Comment:
This was another tweak suggested in the simplification POC, but it's both a bit simpler and also a bit more like life to drive the current values from the historical record, than the historical record from the current values.

Then we can do something like this:

```python
        file = factories.File()
        file_service.get.return_value = file

        canvas_api_client.list_files.return_value = [
            {"id": 1, "display_name": "Wrong name", "size": file.size},
            {"id": 2, "display_name": file.name, "size": 99999999999},
            {"id": 3, "display_name": file.name, "size": file.size},
        ]
```

And demonstrate a number of close matches and not close matches etc.

Reply:
Done:

```python
    def test_find_matching_file_in_course_returns_the_matching_file_id(
        self, finder, canvas_api_client, file_service
    ):
        file_service.get.return_value = factories.File()
        canvas_api_client.list_files.return_value = [
            {"id": 1, "display_name": "File 1", "size": 1024},
            {
                "id": sentinel.matching_file_id,
                "display_name": file_service.get.return_value.name,
                "size": file_service.get.return_value.size,
            },
        ]

        matching_file_id = finder.find_matching_file_in_course(
            sentinel.course_id, sentinel.file_id
        )

        file_service.get.assert_called_once_with(sentinel.file_id, type_="canvas_file")
        canvas_api_client.list_files.assert_called_once_with(sentinel.course_id)
        assert matching_file_id == str(sentinel.matching_file_id)
```